web/alt_text_manager.js

```diff
@@ -0,0 +1,172 @@
+
+    dialog.addEventListener("close", this.#close.bind(this));
+    cancelButton.addEventListener("click", this.#finish.bind(this));
+    saveButton.addEventListener("click", this.#save.bind(this));
+    optionDescription.addEventListener(
+      "change",
+      this.#boundUpdateSaveButtonState
+    );
+    optionDecorative.addEventListener(
+      "change",
+      this.#boundUpdateSaveButtonState
+    );
+    textarea.addEventListener("input", this.#boundUpdateSaveButtonState);
+
+    this.#overlayManager.register(dialog);
+  }
+
+  async editAltText(uiManager, editor) {
+    if (this.#currentEditor || !editor) {
+      return;
+    }
+    const { altText, decorative } = editor.altTextData;
+    if (decorative === true) {
+      this.#optionDecorative.checked = true;
+      this.#optionDescription.checked = false;
+    } else {
+      this.#optionDecorative.checked = false;
+      this.#optionDescription.checked = true;
+    }
+    this.#textarea.value = altText?.trim() || "";
+    this.#updateSaveButtonState();
+
+    this.#currentEditor = editor;
+    this.#uiManager = uiManager;
+    this.#uiManager.removeKeyboardManager();
+
+    const dialog = this.#dialog;
+    await this.#overlayManager.open(dialog);
+
+    const { style } = dialog;
+    const { innerWidth: windowW, innerHeight: windowH } = window;
+    const { width: dialogW, height: dialogH } = dialog.getBoundingClientRect();
+    const { x, y, width, height } = editor.getClientDimensions();
+    const MARGIN = 10;
+    const { direction } = this.#uiManager;
+
+    let left = null;
+    let top = Math.max(0, y - MARGIN);
+    top += Math.min(windowH - (top + dialogH), 0);
+
+    if (direction === "ltr") {
+      // Prefer to position the dialog "after" (so on the right) the editor.
+      if (x + width + MARGIN + dialogW < windowW) {
+        left = x + width + MARGIN;
+      } else if (x > dialogW + MARGIN) {
+        left = x - dialogW - MARGIN;
+      }
+    } else if (x > dialogW + MARGIN) {
+      left = x - dialogW - MARGIN;
+    } else if (x + width + MARGIN + dialogW < windowW) {
+      left = x + width + MARGIN;
+    }
+
+    if (left === null) {
+      top = null;
+      left = Math.max(0, x - MARGIN);
+      left += Math.min(windowW - (left + dialogW), 0);
+      if (y > dialogH + MARGIN) {
+        top = y - dialogH - MARGIN;
+      } else if (y + height + MARGIN + dialogH < windowH) {
+        top = y + height + MARGIN;
+      }
+    }
+
+    if (top !== null) {
+      dialog.classList.add("positioned");
+      style.left = `${left}px`;
+      style.top = `${top}px`;
+    } else {
+      dialog.classList.remove("positioned");
+      style.left = "";
+      style.top = "";
+    }
+  }
+
+  #finish() {
+    if (this.#dialog) {
+      this.#overlayManager.close(this.#dialog);
+    }
+  }
+
+  #close() {
+    this.#uiManager?.addKeyboardManager();
+    this.#currentEditor = null;
+    this.#uiManager = null;
+  }
+
+  #updateSaveButtonState() {
```

Comment:
Nit: Given that it updates both the textarea *and* the button, how about a slightly less specific name?
```suggestion
  #updateUIState() {
```